src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkWriter.java

```diff
@@ -87,37 +125,61 @@ protected void submitRequestEntries(
-
-                // TODO: Make `HttpSinkInternal` retry the failed requests. Currently,
-                //  it does not retry those at all, only adds their count to the
-                //  `numRecordsSendErrors` metric. It is due to the fact we do not have
-                //  a clear image how we want to do it, so it would be both efficient and correct.
-
-                //requestResult.accept(response.getFailedRequests());
-                //} else {
-                //requestResult.accept(Collections.emptyList());
-                //}
+                handleFullyFailedRequest(err, requestEntries, requestResult);
+            } else if (response.getRequests().stream().anyMatch(r -> !r.isSuccessful())) {
+                handlePartiallyFailedRequest(response, requestEntries, requestResult);
+            } else {
+                requestResult.accept(Collections.emptyList());
             }
-            requestResult.accept(Collections.emptyList());
         }, sinkWriterThreadPool);
     }
 
+    private void handleFullyFailedRequest(Throwable err,
+                                          List<HttpSinkRequestEntry> requestEntries,
+                                          Consumer<List<HttpSinkRequestEntry>> requestResult) {
+        int failedRequestsNumber = requestEntries.size();
+        log.error(
+                "Http Sink fatally failed to write all {} requests",
+                failedRequestsNumber);
+        numRecordsSendErrorsCounter.inc(failedRequestsNumber);
+
+        if (deliveryGuarantee == DeliveryGuarantee.AT_LEAST_ONCE) {
+            // Retry all requests.
+            requestResult.accept(requestEntries);
+        } else if (deliveryGuarantee == DeliveryGuarantee.NONE) {
+            // Do not retry failed requests.
+            requestResult.accept(Collections.emptyList());
+        }
+    }
+
+    private void handlePartiallyFailedRequest(SinkHttpClientResponse response,
+                                              List<HttpSinkRequestEntry> requestEntries,
+                                              Consumer<List<HttpSinkRequestEntry>> requestResult) {
+        long failedRequestsNumber = response.getRequests().stream()
+                .filter(r -> !r.isSuccessful())
+                .count();
+        log.error("Http Sink failed to write and will retry {} requests",
+                failedRequestsNumber);
+        numRecordsSendErrorsCounter.inc(failedRequestsNumber);
+
+        if (deliveryGuarantee == DeliveryGuarantee.AT_LEAST_ONCE) {
+            // Assumption: the order of response.requests is the same as requestEntries.
+            // See com.getindata.connectors.http.internal.sink.httpclient.
+            // JavaNetSinkHttpClient#putRequests where requests are submitted sequentially and
+            // then their futures are joined sequentially too.
+            List<HttpSinkRequestEntry> failedRequestEntries = new ArrayList<>();
```

Comment:
Can we have multiple values for the same key in one list of `requestEntries`? It seems to me that by introducing retry mechanism we can interfere with final entries ordering. That may be the case, when latest key entry is saved successfully and then older is retried, effectively overwriting latest value. Do I read that correctly?